### PR TITLE
feat: Set values from local files

### LINKED ISSUE #1754: proposal: import file(s) during `helm install`
More of a feature request as I haven't fleshed out all the details yet.

Right now the only way I can configure config maps and secrets from files is to copy entire file contents (and base 64 for secrets) into the values files or --set, its a bit awkward.

Lets say I have an RSA key, at the moment I base64 this, strip whitespace and put it into my values file:

```yaml
rsaPrivateKey: "LS0tLS1CRUdJTiBSU0EgUFJJVkFURSBLRVktLS0tLQpNSUlFb3dJQkFBS0NBUUVBdUcydDRBZ2dydXl3WjFKa2o4azBWZ1c3ak1vTlhoRVVMSG1mbEpWRFFJbmZTRVBtCnZaWGlYUHRaOVA0VGdwLzQzYmY1MXIvSGt6NVhnbjBlR2FkQlJlZUIxSUxyRXI3N2M5ZXMyenJCWmJUcWZ3UHoKNVNZNzdraXBZVURkczVEOVlsWnUyQk5sMFJzUGpUWWJQOCtvbnNXeVVyNXc2L0cxNEZTK0dnWTA2bGpyTVdPbwpjOVA5dWVqbXlZYitNMUgrMGR4SWl4Wkg5TndBcnJ1RVZJVHlxVGJnaE9tYUhPK1lGUWlySG5XNjk2U01EbmozCnVja21tRTlwS3kwQkk0eUlHdnRIdWJRYTh3U21TMkw0RTA3Y1pXdzkrMW5LdU00Z2E1bStVL3U1WTJSOGJzSm4KVjg5a2ZFVFlVUUF3YUxVVTZySE44UFBlUFZvc3dWWTNjd1kyRXdJREFRQUJBb0lCQUVCWnJLaVZ6V21ibG5XRwpyeVhvYTRWMFhKdGtTTVBBcTh4VTU1UjlmRWpCa3BIeGQwcEJTUWx1ZmxYWWJWQlBweE1keFdQajZLdjh2YVgvCktBaUZibVlJd0NrNVFQSjZodHhGcmIrVXdiUk9LdUFFcmc0UGlDdTlXTFJoWUxudDVpdHd3eDVMcm1RRzBDZFMKY0xHVU1GNnQxbW0rbjBJRVJaU3ZuT0VSdDFibzlVZUE4dlduRjJPU2ZpM2pIWi8zdWFuZmdwcVh1eFVtN202YgplU3cwTDYxVGpLRFU1THBKNDJFcjlJaGRpbG83eXdoREszbTN0ZHNlcS9FY3ltL1I4ZGkwRDFUREEvdTFib3FpCmdRMVZSeFhhVmJQZjhocm5DaVNWWHRoZ01hZTNnTzFqUml0SUlIMlF0a2JUWXJqdzB1cVdIUkVtRVl3dER0aVkKd3NFQ2p3a0NnWUVBOEY0ZFY0cWFXakJlc1kxUm15OU9xdjYyOUpaZS96QjBhRkRESlA5bUhZNlA4NC90YmRKUgpFcFR6eGRXblRjTi83L1pwSDE0V2RFRWVsNUZCaWdRTGQwVlZ4ZUdYd3l1MWN5VTRkMkFCbE56ZU1hWElCK2c1Cm9ZWjA4cWNsQzBmb2NzMVpFOWdjYkpiQnNCbm1XMDh1SlBVVFM4V1gwREpTbFVZTHZScy9ETGNDZ1lFQXhHdzcKVkJLUmY0RDNOOWJqa1JVM1Q0MS9xdXM2Mm9IY1E4VXpzclRoUXFTa0psUDFoRTVPaGtXcFZmTHpJNEltT3VXYQpYYTdKV1doMWZUOS8xd2xWWFJFQkhlM2tnL0pUNlFDeC9YNnJiL1V0U3hTeWdkRTZUbzc5dXhrU3RWRnJ6Q0w3CmFydlk4ZVk4M1VlcGo2UW1tcitYUjhBQjcrY25UckhJUUhHclBZVUNnWUJTSnVpQ1pnWFJNMXo1OUNIakZ5MGMKOWtlVG10MUcvTW1jYVpYNHpDMmJob0VlcnA5b2ZJeG0vRy9BdXdBZWJPWUtIcjdaRUxBOTlZKytIOUExbXFKTQpsbTcxWGVBZnZZMmJtQ29QYXNtL2xpbCs4SGxJY285WDdIdUYvNmlDallZZHpwOVdUL09QRFl1UFk1cSt0SFpCCi9WMGw5WmVvZWJxTitFM0V5VTRvU3dLQmdGUStSMFBMb3ZPWWtHODJqdXVPdWVZNHZlVEVkcjE0ck11MExIUWMKOSsvcFJTM2JISlhlTC9lT3NRM1g0VWY5YmdBY3NaUTBmczhGNmo0Sm1mN1VWZlJRY3JZdmJhN1pvL3R1L09mSQpDQ2RBZEdpYk8rUkpwRVJGTjdXakQwOE0yTHI0MWZQMUZScktzSmdIUUVZSEZUZVJySVJ0YVdkSE1LaWp2MHRFCmRUa3hBb0dCQUlKcE9mTkJZYm9pQlJBL2ZORU1HakNzNUo4SzVlOWlVMG44a1N3NXJlSVNySG5WYy9HckhZSW8KWkxsdDVVdk01bDJ5TGFZRVBRUWpOeGErd05SeUVaemhuZlhGV05kSndRNFdYSVd5WUNmU0FFSGh0UTRnbUZ1RgphczNUZHNqNjU2VGp4dlg0MVU3T21OalVZc0FmYVlQTFN5bzAxM1FjR0JUelBYYlo3R0tUCi0tLS0tRU5EIFJTQSBQUklWQVRFIEtFWS0tLS0tCg=="
```
I know we have access to `(.Files.Glob "foo/*").AsConfig`  and `(.Files.Glob "bar/*").AsSecrets`, but this is restricted to files inside the chart. 

Perhaps we could use go template engine on the value file itself, like:

```yaml
rsaPrivateKey: {{ (.Files.Glob "*.key.pem").AsSecrets }}
```

This would give helm more flexibility, but also more complexity. The functionality could more restricted than whats in chart template files, perhaps even with a different syntax/engine. I'm not sure how it would work with relative paths, that is something we would have to figure out.

I have a feeling this was what #1662 was trying to get across, but I'm not sure. Has anyone else felt like this is needed?

P.S I am so happy with helm, its a great tool and I think you guys are doing a great job.

### PR DESCRIPTION
**Pull request by mumoshu**
This adds the `--set-file key=filepath` flag to `install`, `upgrade`, `template` and `lint` sub-commands so that the content of the file at the `filepath` is set to the value for the `key`.

Resolves #1754

### CODE REVIEW

**Review by cwedgwood — COMMENTED** [review]
lgtm, i will try this tomorrow

**Comment by mumoshu** [comment]
Hi @technosophos @bacongobbler @adamreese 

Would you mind reviewing this? This is an implementation for the feature requested since one and a half years 😉

**Comment by mumoshu** [comment]
@cwedgwood Thanks for taking your time!

**Comment by mumoshu** [comment]
Fixed bugs and tests after a self-review.

**Comment by mumoshu** [comment]
Just realized that I have to also cover `helm lint`. Just a moment...

**Comment by mumoshu** [comment]
Done.

**Comment by mumoshu** [comment]
Missed doc regen 😉Done.

**Comment by davidkarlsen** [comment]
@cwedgwood Any progress on this?

**Comment by davidkarlsen** [comment]
@cwedgwood @technosophos can you please have a look at the PR?

**Comment by cwedgwood** [comment]
@davidkarlsen it's not my call ... given there are nearly 100 merge requests pending, I imagine this might take some time.

**Comment by davidkarlsen** [comment]
@cwedgwood OK - sorry for the nagging - who would be the most appropriate reviewer/merger?

**Comment by cwedgwood** [comment]
@davidkarlsen i think what you've done is fine/appropriate; given the backlog of issues and pull requests you might want to consider forking for now (or merging into another fork that suits you better)

**Comment by davidkarlsen** [comment]
I'm just hunting for the feature and there seems to be a fair share of interest for it, it is actually @mumoshu doing the hard work :)

OK, we'll sit in line and watch the queue :)

**Review by bacongobbler — CHANGES_REQUESTED** [review]
Hey @mumoshu! I had a look at this and the code looks great. Would you mind adding a note about this new feature flag under the following sections?

https://github.com/kubernetes/helm/blob/48e703997016f3edeb4f0b90e6cfdb3456ce3db0/docs/chart_best_practices/values.md#consider-how-users-will-use-your-values

https://github.com/kubernetes/helm/blob/bae4f4bb248a114a0d6cd4fec62d98f5f5453ded/docs/using_helm.md#customizing-the-chart-before-installing

If you find any other sections where --set-file might be useful, please also feel free to add it into the docs. We just need to document new features; we've heard from users that they want to make sure all new features coming into Helm are documented with clear usage examples.

Thanks so much!

**Comment by bacongobbler** [comment]
@davidkarlsen you rang? ;)

Since this PR is a `size/L` we'll also need the sign-off from another maintainer.

**Comment by mumoshu** [comment]
@bacongobbler Thanks for the review. Good point! I completed adding docs to everywhere I think good.

Would you mind taking a look once again?

**Comment by mumoshu** [comment]
Just to make it sure - please feel free to squash everything into a single commit on github when you merge. Of course in any case you want me to do so, just tell me so! Thanks

**Review by bacongobbler — APPROVED** [review]
All very clear examples. Nice touch on adding it to `helm help install`. Thanks!

**Comment by davidkarlsen** [comment]
We need another reviewer, maybe @technosophos ?

**Comment by mumoshu** [comment]
@bacongobbler Would you mind poking me when you're about to merge this?
I'll try to rebase this asap then :)

**Comment by dheerajjoshim** [comment]
This isn't available in 2.10.0-rc.1 ?
https://github.com/kubernetes/helm/releases/tag/v2.10.0-rc.1

**Comment by mumoshu** [comment]
@adamreese Hi! Would it be possible to ask you to merge this?

**Comment by mumoshu** [comment]
Rebased and squashed commits.

File: cmd/helm/lint.go
Comment (technosophos) [inline cmd/helm/lint.go]: Why were TLS options added to Lint, which is explicitly forbidden from making server connections?

Comment (mumoshu) [reply cmd/helm/lint.go]: True! I don't remember why I did it but I'll remove it asap. Thx.

Comment (mumoshu) [reply cmd/helm/lint.go]: Turned out it was due to my misunderstanding. Reverted the tls options and the relevant refactoring altogether - It should be good to go now!

The longer explanation follows.

Firstly, I removed the `vals` func dedicated to the `lint` command and migrated it to reuse another `vals` func used across other commands like `upgrade`.

While migrating,  I believe that I thought it would be better to add support for tls opts for the `vals` func, for use when downloading remote value files. And it resulted in adding those flags. But as you pointed out, it conflicts with the design of the `lint` command, that is forbidden from making server connections. I reverted those changes and it should be good now :)

**Comment by mumoshu** [comment]
@technosophos Thanks for the review! I reverted the changes that turned out wrong. It should be good to go now.

**Comment by mumoshu** [comment]
For your reference, [here's the commits I made](https://github.com/mumoshu/helm/compare/daa0d07...set-value-from-local-file-before-squash) in order to revert things. They're already incorporated and squashed into this PR, but I thought it would help your review.

**Review by technosophos — APPROVED** [review]
Thank you for the quick turn-around on the TLS options. This is looking great.

**Comment by wpc009** [comment]
The [document](https://docs.helm.sh/chart_best_practices/values/#consider-how-users-will-use-your-values) is not very clear about how this `--set-file` option works. 
Finally I figured it out that `--set-file key=path_to_file` will  treat the content of `path_to_file` as multilines string, and set it to `key`. This description should be added in the docs.
The first time I see the `--set-file` option, I expect that helm will import values from the file provided.

**Comment by mumoshu** [comment]
@wpc009 Hey! Thx for the feedback.
Honestly I did my best to make it as clear as possible.
So would you mind proposing alternative sentence(s)?

**Pull request merged**
